### PR TITLE
SPU2: Improve ADMA behaviour/timing

### LINKED ISSUE #2056: Splinter Cell : Audio issue (spu2x)
***PCSX2 version:***
Last build version

***PCSX2 options:***
All default, speedhacks disabled

***Plugins used:***
All default, all core audio tested

***Description of the issue:***
This game have a sound issue related to the spu2x plugin, A kind of cracking is heard when any sort of sound is play.
There is no sound in FMV.

***How to reproduce the issue:***
Load PCSX2 with the spu2x plugin and load the Splinter Cell game.

***Last known version to work:***
Don't exist, this issue is also on the 1.4 version

***PC specifications:***
Intel Core 2 Duo E8600 Gigabyte GTX 780Ti Windows 10 64 Bit

### LINKED ISSUE #2341: Burnout 3 (USA) - Crackling Engine Sounds
***PCSX2 version:***
1.5.0-dev-2237

***PCSX2 options:***
***Plugins used:***
All default, no speedhacks, following GSdx settings:
image

***Description of the issue:***
Frequently, your car's engine will have this annoying crackling to it. It seems to happen most often when I go into a race with software mode.
https://www.youtube.com/watch?v=LiTBlEJnd94

The mods aren't the cause of the issue. I get the problem with an unmodified copy, I just happened to have the modded copy selected when I was recording the video.

***How to reproduce the issue:***
Go into a race with software mode enabled.

***Last known version to work:***
Issue has always been present.

***PC specifications:***
CPU: i7-7700K
GPU: GTX 1050
OS: Windows 10 x64
RAM: 8GB

### LINKED ISSUE #3846: Sound clicks in Vexx
***PCSX2 version:***
v1.7.0-dev-465-g2608673ca

***PCSX2 options:***
default audio settings, but changing them doesn't make any difference

***Plugins used:***
SPU2 built into v1.7.0-dev-465-g2608673ca

***Description of the issue:***
Audio pops and clicks

***How to reproduce the issue:***
Start the game and listen. The issue is more prominent during FMV videos, but you can also hear clicks during gameplay.

***Last known version to work:***
`As far as I'm aware this game never had propper sound

***PC specifications:***
I7 8700, MSI Z370-A PRO, integrated motherboard audio

Blockdump with a couple of seconds of intro video
[Vexx (2021-00-06 19-04-38).zip](https://github.com/PCSX2/pcsx2/files/5776829/Vexx.2021-00-06.19-04-38.zip)

### LINKED ISSUE #4160: [BUG] Myst 3: sound glitches.
**Describe the bug**
In the game "Myst 3" the music (and pretty much every sounds) are shifted left and right consistently during gameplay.

https://user-images.githubusercontent.com/26351275/104856803-d0aabb00-5914-11eb-85cf-534f75e7fdef.mp4

**To Reproduce**
Load Myst 3 and listen to the main menu soundtrack.

**Expected behavior**
The sound should not shift left and right.

**Related issue**

https://github.com/PCSX2/pcsx2/issues/2056

**Emulation Settings**
All default.

**System Info (please complete the following information):**
PCSX2 Revision: pcsx2-v1.7.0-dev-877-g6c81bb54f-windows-x86
OS: Windows 10
CPU: I9 9900k
GPU: RTX 3070

**Dumps**
https://drive.google.com/file/d/126wMAj-FyPdmF5t0HHCtiF2ZU6eJH-3Z/view?usp=sharing

### PR DESCRIPTION
**Pull request by refractionpcsx2**
Seems to have a positive impact on The Simpsons (I know I keep saying this)
Hopefully improves the crackling in burnout 3, Fixes #2341
Apparently fixes music looping in Ratatouille
Fixes #3846 Vexx clicking sound (Partial ADMA transfers)
Fixes the Dolby Pro Logic video missing sound in Primal (Partial ADMA transfers)
Splinter Cell panning audio problem and high pitched noises in cutscenes, Fixes #2056
Myst 3 panning audio problem, Fixes #4160
Dynasty Warriors 5 stuttering videos when Dolby is enabled

If you have any other games which had constantly crackly audio or missing audio, especially in videos, this might help.

### CODE REVIEW

**Comment by atomic83GitHub** [comment]
Unfortunately this PR breaks the Silent Hill 2 black ribbon demo.

**Comment by atomic83GitHub** [comment]
Well, the very last commit in this PR fixes the regression

**Comment by bositman** [comment]
Doesn't fix video stutter in Dynasty Warriors when setting sound to Dolby pro logic. Log:
IPU1 running when IPU1 DMA disabled! CHCR 1 QWC 0
Stream Buffer 1Bank 32768byte 5banks 163840byte used

**Comment by refractionpcsx2** [comment]
> Doesn't fix video stutter in Dynasty Warriors when setting sound to Dolby pro logic. Log:

Now it does :P

File: pcsx2/SPU2/ReadInput.cpp [hunk 1]
```diff
@@ -61,104 +50,87 @@ StereoOut32 V_Core::ReadInput_HiFi()
...
-					else
-						SPU2interruptDMA7();
-				}
+				DMAICounter = InputDataTransferred * 4;
+				LastClock = *cyclePtr;
 			}
 		}
-		InputPosRead &= 0x1ff;
+		else
+			if ((AutoDMACtrl & (Index + 1)))
```
Comment (lightningterror) [inline pcsx2/SPU2/ReadInput.cpp:82]: Maybe else if instead?

File: pcsx2/SPU2/ReadInput.cpp [hunk 2]
```diff
@@ -168,27 +139,14 @@ StereoOut32 V_Core::ReadInput()
 							ConLog("WARNING: adma buffer didn't finish with a whole block!!\n");
 					}
 				}
-
+
```
Comment (lightningterror) [inline pcsx2/SPU2/ReadInput.cpp:142]: Whitespace.

**Pull request merged**